File: Nymeria/nymeria/triggers/discord_cogs/hooks.py
"""Lifecycle-hook authoring commands: the /hook group.

Each subcommand assembles the backend ``/hook`` flag grammar and forwards it via
``_send_backend_command``. Values are ``shlex.quote``-d because the backend
command dispatcher shlex-splits the line back into tokens, so a multi-word
``--text``/``--reason``/``matcher`` must survive as a single token. Event/action
legality (which action is valid for which event) is validated backend-side, so
the cog lists every choice and lets the backend return a friendly error.
"""

from __future__ import annotations

import shlex

import discord
from discord import app_commands
from discord.ext import commands

from typing import TYPE_CHECKING, Optional

if TYPE_CHECKING:
    from ..discord_bot import NymeriaDiscordBot


_EVENT_CHOICES = [
    app_commands.Choice(name="prompt submitted", value="prompt_submit"),
    app_commands.Choice(name="before a tool runs", value="pre_tool_use"),
    app_commands.Choice(name="after a tool runs", value="post_tool_use"),
    app_commands.Choice(name="turn done", value="done"),
]

_ACTION_CHOICES = [
    app_commands.Choice(name="inject context", value="inject_context"),
    app_commands.Choice(name="block if matches", value="block_if_matches"),
    app_commands.Choice(name="rewrite arg", value="rewrite_arg"),
    app_commands.Choice(name="notify", value="notify"),
    app_commands.Choice(name="create todo", value="create_todo"),
    app_commands.Choice(name="webhook", value="webhook"),
]

_CREATE_SCOPE_CHOICES = [
    app_commands.Choice(name="this thread", value="thread"),
    app_commands.Choice(name="global (all threads)", value="global"),
]

_LIST_SCOPE_CHOICES = [
    app_commands.Choice(name="all", value="all"),
    app_commands.Choice(name="this thread + global", value="thread"),
    app_commands.Choice(name="global only", value="global"),
]


def _opt(flag: str, value: Optional[str]) -> list[str]:
    """A ``--flag value`` pair (quoted), or nothing when the value is empty."""
    if not value:
        return []
    return [flag, shlex.quote(value)]


def _kv(key: str, value: Optional[str]) -> list[str]:
    """A ``key=value`` edit token (value quoted), or nothing when unset."""
    if value is None or value == "":
        return []
    return [f"{key}={shlex.quote(value)}"]


class HooksCog(commands.Cog):
    def __init__(self, bot: NymeriaDiscordBot):
        self.bot = bot

    hook_group = app_commands.Group(
        name="hook", description="Author lifecycle hooks (when/logic/return)"
    )

    @hook_group.command(name="list", description="List lifecycle hooks")
    @app_commands.describe(
        scope="Which hooks to show", enabled_only="Only show enabled hooks"
    )
    @app_commands.choices(scope=_LIST_SCOPE_CHOICES)
    async def cmd_hook_list(
        self,
        interaction: discord.Interaction,
        scope: Optional[app_commands.Choice[str]] = None,
        enabled_only: bool = False,
    ):
        await interaction.response.defer(ephemeral=True)
        parts: list[str] = []
        if scope and scope.value == "thread":
            parts += ["--thread", "current"]
        elif scope and scope.value == "global":
            parts.append("--global")
        if enabled_only:
            parts.append("--enabled-only")
        await self.bot._send_backend_command(
            interaction, "hook list", args=" ".join(parts)
        )

    @hook_group.command(name="create", description="Create a lifecycle hook")
    @app_commands.describe(
        name="Hook name",
        event="When it fires",
        action="What it does",
        text="Injected/notify/todo text (inject_context, notify, create_todo)",
        matcher="Tool filter for tool events, e.g. bash or Edit|Write",
        condition="One condition: 'field operator value' (e.g. command contains rm -rf)",
        reason="Denial message (block_if_matches)",
        set_arg="Argument rewrite 'arg=value' (rewrite_arg)",
        url="Webhook URL (webhook)",
        scope="Thread-scoped (default) or global",
        disabled="Create it disabled",
    )
    @app_commands.choices(
        event=_EVENT_CHOICES, action=_ACTION_CHOICES, scope=_CREATE_SCOPE_CHOICES
    )
    async def cmd_hook_create(
        self,
        interaction: discord.Interaction,
        name: str,
        event: app_commands.Choice[str],
        action: app_commands.Choice[str],
        text: Optional[str] = None,
        matcher: Optional[str] = None,
        condition: Optional[str] = None,
        reason: Optional[str] = None,
        set_arg: Optional[str] = None,
        url: Optional[str] = None,
        scope: Optional[app_commands.Choice[str]] = None,
        disabled: bool = False,
    ):
        await interaction.response.defer(ephemeral=True)
        parts = [shlex.quote(name), "--event", event.value, "--action", action.value]
        parts += _opt("--text", text)
        parts += _opt("--matcher", matcher)
        parts += _opt("--cond", condition)
        parts += _opt("--reason", reason)
        parts += _opt("--set", set_arg)
        parts += _opt("--url", url)
        if scope:
            parts += ["--scope", scope.value]
        if disabled:
            parts.append("--disabled")
        await self.bot._send_backend_command(
            interaction, "hook create", args=" ".join(parts)
        )

    @hook_group.command(name="show", description="Show one hook's configuration")
    @app_commands.describe(hook_id="The hook ID (first 8 chars is enough)")
    async def cmd_hook_show(self, interaction: discord.Interaction, hook_id: str):
        await interaction.response.defer(ephemeral=True)
        await self.bot._send_backend_command(
            interaction, "hook show", args=shlex.quote(hook_id)
        )

    @hook_group.command(name="test", description="Dry-run a hook and preview its output")
    @app_commands.describe(hook_id="The hook ID (first 8 chars is enough)")
    async def cmd_hook_test(self, interaction: discord.Interaction, hook_id: str):
        await interaction.response.defer(ephemeral=True)
        await self.bot._send_backend_command(
            interaction, "hook test", args=shlex.quote(hook_id)
        )

    @hook_group.command(name="enable", description="Enable a hook")
    @app_commands.describe(hook_id="The hook ID (first 8 chars is enough)")
    async def cmd_hook_enable(self, interaction: discord.Interaction, hook_id: str):
        await interaction.response.defer(ephemeral=True)
        await self.bot._send_backend_command(
            interaction, "hook enable", args=shlex.quote(hook_id)
        )

    @hook_group.command(name="disable", description="Disable a hook")
    @app_commands.describe(hook_id="The hook ID (first 8 chars is enough)")
    async def cmd_hook_disable(self, interaction: discord.Interaction, hook_id: str):
        await interaction.response.defer(ephemeral=True)
        await self.bot._send_backend_command(
            interaction, "hook disable", args=shlex.quote(hook_id)
        )

    @hook_group.command(name="delete", description="Delete a hook permanently")
    @app_commands.describe(hook_id="The hook ID (first 8 chars is enough)")
    async def cmd_hook_delete(self, interaction: discord.Interaction, hook_id: str):
        await interaction.response.defer(ephemeral=True)
        await self.bot._send_backend_command(
            interaction, "hook delete", args=f"{shlex.quote(hook_id)} --yes"
        )

    @hook_group.command(
        name="edit", description="Edit a hook's fields, condition, or rewrite"
    )
    @app_commands.describe(
        hook_id="The hook ID (first 8 chars is enough)",
        name="New name",
        enabled="Enable or disable",
        event="New event",
        action="Switch the action",
        matcher="New tool filter, e.g. bash or Edit|Write",
        text="New injected/notify/todo text",
        reason="New denial message",
        url="New webhook URL",
        condition="Replace conditions with one 'field operator value'",
        set_arg="Replace rewrites with one 'arg=value'",
    )
    @app_commands.choices(event=_EVENT_CHOICES, action=_ACTION_CHOICES)
    async def cmd_hook_edit(
        self,
        interaction: discord.Interaction,
        hook_id: str,
        name: Optional[str] = None,
        enabled: Optional[bool] = None,
        event: Optional[app_commands.Choice[str]] = None,
        action: Optional[app_commands.Choice[str]] = None,
        matcher: Optional[str] = None,
        text: Optional[str] = None,
        reason: Optional[str] = None,
        url: Optional[str] = None,
        condition: Optional[str] = None,
        set_arg: Optional[str] = None,
    ):
        await interaction.response.defer(ephemeral=True)
        parts = [shlex.quote(hook_id)]
        parts += _kv("name", name)
        if enabled is not None:
            parts.append(f"enabled={'true' if enabled else 'false'}")
        if event:
            parts.append(f"event={event.value}")
        if action:
            parts.append(f"action={action.value}")
        parts += _kv("matcher", matcher)
        parts += _kv("text", text)
        parts += _kv("reason", reason)
        parts += _kv("url", url)
        parts += _opt("--cond", condition)
        parts += _opt("--set", set_arg)
        await self.bot._send_backend_command(
            interaction, "hook edit", args=" ".join(parts)
        )

File: nymeria-desktop/src/lib/utils/hooks.ts
/**
 * Lifecycle-hook taxonomy: the event/action legality map plus the
 * category-coded families (Guardrails / Context / Reactions) that drive the
 * authoring form and the feed grouping. Mirrors the backend
 * (`core/hook_manager.py` EVENT_ACTIONS + action families). Pure data + pure
 * functions so both apps keep a byte-identical copy.
 */

import type { Hook, HookAction, HookCondition, HookEvent } from '$lib/types';

export type HookCategory = 'guardrails' | 'context' | 'reactions' | 'commands';

/** Which actions are legal for each event (gates the authoring action list). */
export const HOOK_EVENT_ACTIONS: Record<HookEvent, HookAction[]> = {
  prompt_submit: ['inject_context', 'run_command', 'run_workflow'],
  pre_tool_use: [
    'block_if_matches',
    'rewrite_arg',
    'require_approval',
    'run_command',
    'run_workflow',
  ],
  post_tool_use: [
    'inject_context',
    'notify',
    'create_todo',
    'webhook',
    'run_command',
    'run_workflow',
  ],
  done: ['inject_context', 'notify', 'create_todo', 'webhook', 'run_command', 'run_workflow'],
  command_submit: [
    'block_if_matches',
    'rewrite_arg',
    'require_approval',
    'notify',
    'create_todo',
    'webhook',
    'run_command',
    'run_workflow',
  ],
};

/** Events that fire around a tool call, where a tool-name matcher applies. */
export const HOOK_TOOL_EVENTS: HookEvent[] = ['pre_tool_use', 'post_tool_use'];

/** Events that fire around a slash-command dispatch: the matcher targets
 * canonical command paths ('tools list', 'provider *') instead. */
export const HOOK_COMMAND_EVENTS: HookEvent[] = ['command_submit'];

/** All events on which the name matcher applies (the authoring form's gate). */
export const HOOK_MATCHER_EVENTS: HookEvent[] = [...HOOK_TOOL_EVENTS, ...HOOK_COMMAND_EVENTS];

const CATEGORY_OF: Record<HookAction, HookCategory> = {
  block_if_matches: 'guardrails',
  rewrite_arg: 'guardrails',
  require_approval: 'guardrails',
  inject_context: 'context',
  notify: 'reactions',
  create_todo: 'reactions',
  webhook: 'reactions',
  run_command: 'commands',
  run_workflow: 'commands',
};

export function hookCategory(action: HookAction | string): HookCategory {
  // The reserved turn-metadata system hook (action outside the authorable
  // union) injects the built-in [Time:]/[Trigger:] block: a context hook.
  if (action === 'turn_metadata') return 'context';
  return CATEGORY_OF[action as HookAction] ?? 'reactions';
}

export interface HookCategoryMeta {
  key: HookCategory;
  label: string;
  icon: string;
  description: string;
}

/** Ordered for display: guard first (most consequential), then context, reactions. */
export const HOOK_CATEGORIES: HookCategoryMeta[] = [
  {
    key: 'guardrails',
    label: 'Guardrails',
    icon: 'warning',
    description: 'Block or rewrite tool calls before they run',
  },
  {
    key: 'context',
    label: 'Context',
    icon: 'fileText',
    description: 'Add text into the conversation',
  },
  {
    key: 'reactions',
    label: 'Reactions',
    icon: 'bell',
    description: 'Notify, create a task, or call a webhook',
  },
  {
    key: 'commands',
    label: 'Custom logic',
    icon: 'terminal',
    description: 'Run a shell command (admin only) or a nym workflow',
  },
];

export interface HookActionMeta {
  label: string;
  icon: string;
  /** Short helper shown under the action picker. */
  hint: string;
}

export const HOOK_ACTION_META: Record<HookAction, HookActionMeta> = {
  inject_context: {
    label: 'Inject context',
    icon: 'fileText',
    hint: 'Add text (with {placeholders}) into the conversation.',
  },
  block_if_matches: {
    label: 'Block if matches',
    icon: 'warning',
    hint: 'Deny the tool call when the conditions match.',
  },
  rewrite_arg: {
    label: 'Rewrite argument',
    icon: 'edit',
    hint: "Change a tool argument before the call runs.",
  },
  require_approval: {
    label: 'Require approval',
    icon: 'flag',
    hint: 'Hold the tool call until you approve or deny it (no answer = deny).',
  },
  notify: {
    label: 'Notify',
    icon: 'bell',
    hint: 'Send a notification.',
  },
  create_todo: {
    label: 'Create task',
    icon: 'check',
    hint: 'Add a task to your list.',
  },
  webhook: {
    label: 'Webhook',
    icon: 'globe',
    hint: 'POST a rendered body to a URL.',
  },
  run_command: {
    label: 'Run command',
    icon: 'terminal',
    hint: 'Run a shell command (admin only; must be enabled on the server).',
  },
  run_workflow: {
    label: 'Run workflow',
    icon: 'bolt',
    hint: "Run a published, approved nym workflow as this hook's logic.",
  },
};

/**
 * System actions the backend can return but users cannot author (excluded from
 * HOOK_EVENT_ACTIONS deliberately). Today: the reserved `turn-metadata` hook
 * that GET /hooks synthesizes for every user.
 */
const SYSTEM_ACTION_META: Record<string, HookActionMeta> = {
  turn_metadata: {
    label: 'Turn metadata',
    icon: 'clock',
    hint: 'Built-in [Time:]/[Trigger:] stamp injected at the start of each turn.',
  },
};

const UNKNOWN_ACTION_ICON = 'settings';

/**
 * Meta for ANY action the wire can carry. The typed HOOK_ACTION_META table
 * only knows the nine authorable actions, so feed rows must resolve through
 * this accessor: a system or newer-backend action degrades to a labeled row
 * instead of crashing the dashboard on `undefined.icon`.
 */
export function hookActionMeta(action: HookAction | string): HookActionMeta {
  return (
    HOOK_ACTION_META[action as HookAction] ??
    SYSTEM_ACTION_META[action] ?? { label: action, icon: UNKNOWN_ACTION_ICON, hint: '' }
  );
}

export interface HookEventMeta {
  label: string;
  hint: string;
}

export const HOOK_EVENT_META: Record<HookEvent, HookEventMeta> = {
  prompt_submit: {
    label: 'On your message',
    hint: 'Fires when you send a message, before the agent responds.',
  },
  pre_tool_use: {
    label: 'Before a tool runs',
    hint: 'Fires before a tool call; can block or rewrite it.',
  },
  post_tool_use: {
    label: 'After a tool runs',
    hint: 'Fires after a tool call finishes.',
  },
  done: {
    label: 'When the turn ends',
    hint: "Fires once the agent's reply is complete.",
  },
  command_submit: {
    label: 'Before a command runs',
    hint: 'Fires when a slash command is dispatched, on every surface; can block, rewrite, or require approval.',
  },
};

/** Null-safe twin of hookActionMeta for events a newer backend might add. */
export function hookEventMeta(event: HookEvent | string): HookEventMeta {
  return HOOK_EVENT_META[event as HookEvent] ?? { label: event, hint: '' };
}

export const HOOK_OPERATORS: { value: HookCondition['operator']; label: string }[] = [
  { value: 'contains', label: 'contains' },
  { value: 'equals', label: 'equals' },
  { value: 'not_equals', label: 'not equals' },
  { value: 'starts_with', label: 'starts with' },
  { value: 'matches_regex', label: 'matches regex' },
];

/**
 * Operators for the definition-level fire gate. Superset of HOOK_OPERATORS: adds
 * the numeric comparisons (float-coerced backend-side) the context-usage fields
 * (`context_pct_of_trigger`, `context_tokens`, ...) rely on. Matches the backend
 * `ConditionOperator` set (core/conditions.py).
 */
export const FIRE_GATE_OPERATORS: { value: HookCondition['operator']; label: string }[] = [
  { value: 'contains', label: 'contains' },
  { value: 'equals', label: 'equals' },
  { value: 'not_equals', label: 'not equals' },
  { value: 'starts_with', label: 'starts with' },
  { value: 'matches_regex', label: 'matches regex' },
  { value: 'gt', label: '> (greater than)' },
  { value: 'gte', label: '>= (at least)' },
  { value: 'lt', label: '< (less than)' },
  { value: 'lte', label: '<= (at most)' },
];

/**
 * Short human summary of a hook's logic for the feed row. Reads the flat
 * `logic` dict the backend returns (discriminated on `action`). Switches on
 * the widened action so the wire's system action (`turn_metadata`, outside
 * the authorable `HookAction` union) resolves to its case, not the fallback.
 */
export function describeHookLogic(hook: Hook): string {
  const logic = hook.logic || {};
  const matcher = hook.matcher ? ` on ${hook.matcher}` : '';
  switch (hook.action as HookAction | string) {
    case 'inject_context':
    case 'notify':
    case 'create_todo': {
      const text = String(logic.text ?? hook.text ?? '').replace(/\s+/g, ' ').trim();
      return text || '(empty)';
    }
    case 'webhook': {
      const url = String(logic.url ?? '').trim();
      return url ? `POST ${url}` : '(no URL)';
    }
    case 'block_if_matches': {
      const conds = Array.isArray(logic.conditions) ? logic.conditions : [];
      const when = conds.length
        ? (conds as HookCondition[])
            .map((c) => `${c.field} ${c.operator} ${c.value}`)
            .join(' and ')
        : 'always';
      return `Deny${matcher} when ${when}`;
    }
    case 'rewrite_arg': {
      const updates = (logic.updates ?? {}) as Record<string, string>;
      const keys = Object.keys(updates);
      return keys.length ? `Rewrite ${keys.join(', ')}${matcher}` : `Rewrite args${matcher}`;
    }
    case 'require_approval': {
      const conds = Array.isArray(logic.conditions) ? logic.conditions : [];
      const when = conds.length
        ? (conds as HookCondition[])
            .map((c) => `${c.field} ${c.operator} ${c.value}`)
            .join(' and ')
        : 'always';
      const window = Number(logic.timeout_seconds ?? 180);
      return `Hold for approval${matcher} when ${when} (${window}s window)`;
    }
    case 'run_command': {
      const command = String(logic.command ?? '').replace(/\s+/g, ' ').trim();
      return command ? `Run ${command}${matcher}` : '(no command)';
    }
    case 'run_workflow': {
      const workflowId = String(logic.workflow_id ?? '').trim();
      return workflowId ? `Run workflow ${workflowId}${matcher}` : '(no workflow)';
    }
    case 'turn_metadata': {
      // The system turn-metadata hook: show the template it injects. Newlines
      // are kept (the two-line [Time:]/[Trigger:] frame IS the content;
      // desktop renders the summary pre-wrap, mobile's compact row collapses
      // it to one ellipsized line by design).
      const template = String(logic.text ?? '').trim();
      return template || '(built-in template)';
    }
    default:
      return hook.action;
  }
}
